lib/ipmi_delloem.c

```diff
@@ -338,6 +365,689 @@ usage(void)
+
+	if( (iDRAC_FLAG == IDRAC_12G) || (iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G) ) {
+		req.msg.netfn = DELL_OEM_NETFN;
+		req.msg.lun = 0;
+		req.msg.cmd = 0xbf;
+		req.msg.data = data;
+		req.msg.data_len = 1;
+		if((iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G))
+			data[0] = 0x04;
+		else
+			data[0] = 0x01;
+
+		rsp = intf->sendrecv(intf, &req);
+		if (rsp == NULL) {
+				printf(" Error Getting the Firmware version \n");
+				return -1;
+		}
+		if (rsp->ccode == 0xc1) {
+			memset(&firmware_ver,0,sizeof(struct FW_VER));
+		}
+		else if (rsp->ccode > 0) {
+				printf(" Error getting device ID (%s)\n",
+					val2str(rsp->ccode, completion_code_vals));
+				return -1;
+		}
+		else
+			memcpy(&firmware_ver,(struct FW_VER*)rsp->data,sizeof(struct FW_VER));
+	}
+
+	memset(&req, 0, sizeof(req));
+	req.msg.netfn = IPMI_NETFN_APP;
+	req.msg.lun = 0;
+	req.msg.cmd = BMC_GET_DEVICE_ID;
+	req.msg.data = NULL;
+	req.msg.data_len = 0;
+
+	rsp = intf->sendrecv(intf, &req);
+	if (rsp == NULL) {
+		printf(" Error getting device ID (No Response)\n");
+		return -1;
+	}
+	if (rsp->ccode > 0) {
+		printf(" Error getting device ID (%s)\n",
+			val2str(rsp->ccode, completion_code_vals));
+		return -1;
+	}
+
+	dev_id = (struct ipm_devid_rsp *)(void *)rsp->data;
+
+	printf("Device ID           : %0.2d\n", dev_id->device_id);
+	printf("Device Revision     : %0.2d\n", dev_id->device_revision & 0x7F);
+	if( (iDRAC_FLAG == IDRAC_12G) || (iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G)) {
+		if(firmware_ver.major_ver == 0) {
+			printf("Firmware Version    : %d.%0.2d(Build %02d)\n", dev_id->fw_rev1 & 0x7F,
+				dev_id->fw_rev2,dev_id->aux_fw_rev[1]);
+		} else {
+			if((iDRAC_FLAG == IDRAC_13G) || (iDRAC_FLAG == IDRAC_14G)) {
+				printf("Firmware Version    : %d.%0.2d.%02d.%02d(Build %02d)\n", firmware_ver.major_ver,
+				firmware_ver.minor_ver,(firmware_ver.pla_data_msb << 8 | firmware_ver.pla_data_lsb),
+				(firmware_ver.lc_ver_msb << 8 | firmware_ver.lc_ver_lsb),
+				(firmware_ver.build_no_msb << 8 | firmware_ver.build_no_lsb));
+			} else {
+				printf("Firmware Version    : %d.%0.2d.%02d(Build %02d)\n", firmware_ver.major_ver,
+				firmware_ver.minor_ver,(firmware_ver.pla_data_msb << 8 | firmware_ver.pla_data_lsb),
+				(firmware_ver.build_no_msb << 8 | firmware_ver.build_no_lsb));
+			}
+		}
+	}else {
+		// DF427387 Fix for Build Version Number
+		//	Byte:	  13				14		   15	       16
+		//	Data: <platform ID> <Build Number> <RES> <RES>
+		printf("Firmware Version    : %d.%0.2x.%02d\n", dev_id->fw_rev1 & 0x7F,
+			dev_id->fw_rev2,dev_id->aux_fw_rev[1]);
+	}
+
+	/* BCD value. byte->hex conversion used.*/
+	printf("IPMI Version        : %d.%d\n", dev_id->ipmi_version & 0x0F, dev_id->ipmi_version >> 4);
+	ipmi_version = dev_id->ipmi_version;
+
+
+	printf("Manufacturer ID     : %lu\n",
+		(uint32_t)(dev_id->manufacturer_id[2] & 0x0f) * 65536 +
+		(uint32_t)(dev_id->manufacturer_id[1] * 256) +
+		(uint32_t)dev_id->manufacturer_id[0]);
```

Comment:
There is `IPMI_DEV_MANUFACTURER_ID()` for that. But yet again, I don't like the very idea of this function as it seems to copy the functionality of `ipmitool mc info`, `ipmitool fru print`, and/or `ipmitool ekanalyzer frushow`.